mitxgraders/helpers/math_array.py

```diff
@@ -0,0 +1,371 @@
+from __future__ import division # necessary for one of the doctests
+import numpy as np
+from numbers import Number
+from mitxgraders.baseclasses import StudentFacingError
+from mitxgraders.helpers.mathfunc import robust_pow
+
+class MathArrayError(StudentFacingError):
+    """
+    Thrown by MathArray when anticipated errors are made.
+    """
+
+def is_number_zero(value):
+    """
+    Tests whether a value is the scalar number 0.
+
+    >>> map(is_number_zero, [0, 0.0, 0j])
+    [True, True, True]
+    >>> is_number_zero(np.matrix([0, 0, 0]))
+    False
+    """
+    return isinstance(value, Number) and value == 0
+
+def is_scalar_array_zero(value):
+    return isinstance(value, MathArray) and value.ndim == 0 and value.item() == 0
+
+def is_square(array):
+    return array.ndim == 2 and array.shape[0] == array.shape[1]
+
+class MathArray(np.ndarray):
+    """
+    A modification of numpy's ndarray class that behaves more like a mathematician would expect.
+
+    Terminology:
+    ============
+        - vector, 1 dimensional, e.g., MathArray([1, 2, 3])
+        - matrix, 2 dimensional, e.g., MathArray([[1, 2], [3, 4]])
+        - tensor, 3+ dimensional, e.g., MathArray([ [[1]], [[2]] ])
+
+    Zero-dimensional arrays, e.g., MathArray(5) are also supported but should
+    not occur in practice.
+
+        - scalar: Python number OR zero-dimensional array
+
+    Differences from numpy.ndarray:
+    ==============================
+        - Multiplication:
+            - scalar * array = scales elements of array;
+            - array * scalar = scales elements of array;
+            If shapes are correct:
+            - vector * vector = number
+            - matrix * vector = vector
+            - vector * vector = vector
+            - matrix * matrix = matrix
+            Any multiplication involving tensors and arrays of dimension >1 is not supported.
+        - powers are only allowed for square matrices and integer-like exponents.
+        - supports multiplication with a universal identity irrespective of own shape
+        - supports addition/subtraction with a universal identity if MathArray is square matrix.
+        - throws MathArrayError instances when anticipated errors are made.
+        Raised with student-friendly error messages.
+    """
+
+    def __new__(cls, input_array):
+        """
+        Creates a new MathArray Instance.
+
+        Note: The ndarray constructor is very low-level and apparently not meant
+        to be used, hence we need our own. This is a quirk of subclassing
+        ndarray. See https://docs.scipy.org/doc/numpy/user/basics.subclassing.html
+        for more info.
+        """
```

Comment:
Add a comment that `array_finalize` is unnecessary, as we don't add any properties to the object.